src/Datadog.Trace/Plugins/PluginManager.cs

```diff
@@ -0,0 +1,117 @@
+
+                return Loaded;
+            }
+
+            var loadedPlugins = TryLoadPlugins(pluginFiles);
+
+            Log.Information("Successfully loaded '{0}' plugin(s).", property: loadedPlugins.Count);
+
+            return Loaded = loadedPlugins;
+        }
+
+        private static IReadOnlyCollection<IOTelPlugin> TryLoadPlugins(string[] pluginFiles)
+        {
+            var loaded = new List<IOTelPlugin>();
+
+            foreach (string file in pluginFiles)
+            {
+                string fullPath = Path.GetFullPath(file);
+
+                if (File.Exists(fullPath))
+                {
+                    try
+                    {
+                        Assembly pluginAssembly = Assembly.LoadFrom(fullPath);
+                        IOTelPlugin plugin = ConvertToPlugin(pluginAssembly);
+
+                        if (plugin != null)
+                        {
+                            loaded.Add(plugin);
+
+                            Log.Information("Plugin assembly loaded '{0}'.", pluginAssembly.FullName);
+                        }
+                        else
+                        {
+                            Log.Warning("Could not load {0} from '{1}'.", nameof(IOTelPlugin), pluginAssembly.FullName);
+                        }
+                    }
+                    catch (Exception ex) when (
+                        ExceptionUtil.IsAssemblyLoadException(ex) ||
+                        ExceptionUtil.IsDynamicInvocationException(ex))
+                    {
+                        Log.Warning(ex, "Plugin assembly could not be loaded. Skipping vendor plugin load.");
+                    }
+                }
+                else
+                {
+                    Log.Warning("Plugin path is defined but could not find the path '{0}'.", fullPath);
+                }
+            }
+
+            return loaded;
+        }
+
+        private static IOTelPlugin ConvertToPlugin(Assembly assembly)
+        {
+            var pluginType = typeof(IOTelPlugin);
+
+            var pluginTypes = assembly
+                .GetTypes()
+                .Where(p => pluginType.IsAssignableFrom(p))
+                .ToList();
+
+            var pluginInstance = pluginTypes
+                .Select(p => (IOTelPlugin)Activator.CreateInstance(p))
+                .FirstOrDefault();
+
+            if (pluginTypes.Count > 1)
+            {
+                Log.Warning("Detected {0} plugins in the assembly '{1}'. Loading only the first type '{2}'.", pluginTypes.Count, assembly.FullName, pluginInstance.GetType().FullName);
+            }
+
+            return pluginInstance;
+        }
+
+        private static string GetPluginsRuntimeName()
+        {
+            // returns the runtime directory of the current running tracer
+            return Directory.GetParent(typeof(PluginManager).Assembly.Location).Name;
```

Comment:
This can cause NullReferenceException in theory, I would change it to:
```csharp
return Directory.GetParent(typeof(PluginManager).Assembly.Location)?.Name;
```
so it's all safe

Reply:
Well in theory if `?.Name` returns `null` then it's a way bigger problem 😄 One option is to convert `NullReferenceException` to `InvalidOperationException` etc, but for me it doesn't give much.

Reply:
This doesn't seem like the best way to get the name of the runtime, especially if running in the GAC will be a scenario. What about using the `FrameworkDescription.Instance` to get the name?

Reply:
Any reason why not getting the fully qualified assembly name:

```csharp
return typeof(PluginManager).Assembly.FullName.ToString()
```
?

Reply:
@zacharycmontoya  Is it possible to run the tracer from the GAC? Any docs?

Reply:
Had to rework it completely. Should be more proper solution for finding out the target framework. Thanks @zacharycmontoya for the direction!